### ISSUE TITLE
Question: is it possible to use a custom driver?

### ISSUE THREAD

**Issue by goofy2k** [thestk/rtmidi#274]
I am working on a project where I want to use BLE for MIDI output. I have this running for sending single commands. Now I want to use a library that uses RtMidi for input/output.  Is there a way (preferably via an API) to bind a custom driver to RtMidi?   I read in the docs about openVirtualPort, but that is probably not what I look for, as I don't see a way to e.g. give a pointer to a callback/custom driver.

**Comment by keinstein** [comment]
Yes, it is. You just have to write the glue code that supports the RtMidiApi virtual functions.

If you want to make a Operating System driver which fits directly to Windows (probably a fake USB device, ALSA (RAW interface), macOS (OSS), or whatever, you should write a driver for each system separately as you have to deal with BLE also on the OS level.

If you want to make a quick hack, openVirtualPort is exactly what you want (unfortunately it does not work with WinMM).

**Comment by goofy2k** [comment]
OK, thanks! With these insights, I will give it another try. Do you
know/have an example?

Fred

Op zo 21 nov. 2021 16:21 schreef Tobias Schlemmer ***@***.***

**Comment by goofy2k** [comment]
I had a closer look at the API. I did not find information on how to create a procedure/driver that will be detected as an available input or output port when scanning for them.....

**Comment by insolace** [comment]
Use virtualports and address them from your other applications by name.
